packages/controllers/src/secure_admin.rs

```diff
@@ -0,0 +1,1091 @@
+use std::fmt::Debug;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{
+    Addr, Api, CustomQuery, DepsMut, MessageInfo, Response, StdError, StdResult, Storage,
+};
+use cw_storage_plus::Item;
+use schemars::JsonSchema;
+use thiserror::Error;
+
+/// Returned from Admin.query()
+#[cw_serde]
+pub struct SecureAdminResponse {
+    pub admin: Option<String>,
+    pub proposed: Option<String>,
+}
+
+/// Errors returned from Admin state transitions
+#[derive(Error, Debug, PartialEq)]
+pub enum SecureAdminError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("Caller is not admin")]
+    NotAdmin {},
+
+    #[error("Caller is not the proposed admin")]
+    NotProposedAdmin {},
+
+    #[error("Admin state transition was not valid")]
+    StateTransitionError {},
+}
+
+type AdminResult<T> = Result<T, SecureAdminError>;
+
+/// The finite states that are possible
+#[cw_serde]
+enum AdminState {
+    A(AdminUninitialized),
+    B(AdminSetNoneProposed),
+    C(AdminSetWithProposed),
+    D(AdminRoleAbolished),
```

Comment:
any particular reason to give these single character names? imo, it makes the code a little hard to follow elsewhere because you've got to map back in your mind.

Reply:
I found AdminUninitialized(AdminUninitialized) to be a bit verbose when my IDE gives me the preview:
<img width="424" alt="Screenshot 2022-12-07 at 8 57 58 AM" src="https://user-images.githubusercontent.com/16624263/206121426-0e4614c4-b6f0-4ee7-b29b-470bdffd8ded.png">


But perhaps that's better having it like that 🤷